KAFKA-5588: Remove deprecated new-consumer option for tools

ppatierno opened the pull request:
Removed the deprecated "--new-consumer" option for all consumer based tools.
This option was already deprecated in the previous version so it can be removed in the coming new 2.0.0 major version.

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

ppatierno committed:
Removed the deprecated "--new-consumer" option for all consumer based tools

viktorsomogyi reviewed (commented):
Reviewed where do we use the new-consumer option. It seems that inside the ducktapes we still use them. Do you think those usages should be removed from there too?
https://github.com/apache/kafka/blob/d99f4a0ffa65ae3a654b043547f114a16fbe4905/tests/kafkatest/services/console_consumer.py#L195
https://github.com/apache/kafka/blob/d99f4a0ffa65ae3a654b043547f114a16fbe4905/tests/kafkatest/services/kafka/kafka.py#L587
https://github.com/apache/kafka/blob/d99f4a0ffa65ae3a654b043547f114a16fbe4905/tests/kafkatest/services/kafka/kafka.py#L614
https://github.com/apache/kafka/blob/d99f4a0ffa65ae3a654b043547f114a16fbe4905/tests/kafkatest/services/performance/consumer_performance.py#L48
https://github.com/apache/kafka/blob/d99f4a0ffa65ae3a654b043547f114a16fbe4905/tests/kafkatest/services/performance/consumer_performance.py#L109

ppatierno commented:
@viktorsomogyi I was thinking the same thing but first waiting for a review by some committer.

vahidhashemian reviewed (commented):
Thank you for the patch.

Would it be possible to also add a test case for both `ConsumerGroupCommand` and `ConsoleConsumer` that verifies using the `--new-consumer` option causes an error?

I think the upgrade document also needs to be updated to mention this change.

ppatierno commented:
@vahidhashemian thanks for the feedback.
Regarding `ConsoleConsumer`, the `ConsumerConfig` class uses the `tryParse` method which just prints the help and exits if a wrong option is specified. It's different from the `ConsumerPerformace` where an exception is raised in this case (so I use that as "expected" for the unit test). Any opinion on that?
Regarding the `ConsumerGroupCommand`, we have the `ConsumerGroupCommandTest` class which is just base class for other tests but it doesn't have unit tests aboud configu itself.
I was thinking to not modify the current `ConsumerGroupCommandTest` in the `kafka.admin` package but adding a `ConsumerGroupCommandTest` in the `kafka.tools` package alongside `ConsoleConsumerTest` and `ConsumerPerformanceTest`. Wdyt ?
Regarding the upgrade doc I'll update that, thanks!

ppatierno committed:
Updated upgrade doc with description of KIP-176

vahidhashemian commented:
@ppatierno, do you think for `ConsoleConsumerTest` we can do something like in `shouldExitOnInvalidConfigWithAutoOffsetResetAndConflictingFromBeginningOldConsumer`? Also, for consumer groups, maybe something similar to `testDescribeWithMultipleSubActions` in `DescribeConsumerGroupTest`?

ijuma reviewed (commented):
Thanks for the PR. Please search the *.py code too as it has a few mentions of `--new-consumer`.

ppatierno commented:
@viktorsomogyi @ijuma related to the Python files as far as I can see the `--new-consumer` option is needed for Kafka version <= 0.10.0.0 and there is a check for that in the console_consumer.py and consumer_performance.py files.
The same doesn't happen in the kafka.py.
Instead of removing the option from there I think that such a check should be added to kafka.py as well.
Wdyt ?

ijuma commented:
@ppatierno Sounds good.

ppatierno committed:
Added check on using --new-consumer based on the node version

ppatierno committed:
Added unit tests for checking unrecognized --new-consumer option

ppatierno commented:
@vahidhashemian thanks for the suggestions, I didn't notice that `Exit` class provided a way for changing the exit behaviour. I added the unit tests which make sense about unrecognized `--new-consumer` option; I added the test for all the possible consumer group classes (describe, list, delete and reset) for consistency because they are different mutual options for the same tool.

@viktorsomogyi @ijuma I updated the Python tests as proposed.

vahidhashemian reviewed (approved):
Thanks for addressing my comment. LGTM.

viktorsomogyi reviewed (approved):
Thanks for addressing my comments too.

ppatierno commented:
@vahidhashemian @viktorsomogyi thanks for your review and for accepting the changes.
@ijuma have you time to review and maybe merged by tomorrow ? Tomorrow is the deadline as far as I understood from Rajini emal : 

> Please ensure that all KIPs for 2.0.0 have been merged by Tuesday the 5th of June. Any remaining KIPs will be moved to the next release

ijuma commented on core/src/test/scala/unit/kafka/admin/DeleteConsumerGroupsTest.scala:
```diff
@@ -222,4 +222,11 @@ class DeleteConsumerGroupTest extends ConsumerGroupCommandTest {
+  def testDeleteWithUnrecognizedNewConsumerOption() {
+    val cgcArgs = Array("--new-consumer", "--bootstrap-server", brokerList, "--delete", "--group", group)
+    getConsumerGroupService(cgcArgs)
+    fail("Expected an error due to presence of unrecognized --new-consumer option")
```
This is not needed given the expected exception annotation. Same for other similar cases.

omkreddy replied:
Do we really need  tests for  checking unrecognized  option "--new-consumer" ?
maybe one round of manual testing should be sufficient.  maintaining these tests in repo may not be worth.

ijuma commented on core/src/test/scala/unit/kafka/admin/DeleteConsumerGroupsTest.scala:
```diff
@@ -222,4 +222,11 @@ class DeleteConsumerGroupTest extends ConsumerGroupCommandTest {
         result.keySet.contains(group) && result.get(group).contains(Errors.COORDINATOR_NOT_AVAILABLE))
   }
+
+  @Test(expected = classOf[joptsimple.OptionException])
```
Nit: why are we not importing `OptionException`?

ijuma reviewed (commented):
Just some minor nits.

ijuma commented on core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:
```diff
@@ -185,6 +183,30 @@ class ConsoleConsumerTest {
+
+    // Override exit procedure to throw an exception instead of exiting, so we can catch the exit
+    // properly for this test case
+    Exit.setExitProcedure((_, message) => throw new IllegalArgumentException(message.orNull))
```
This comment seems redundant given the code.

ppatierno replied:
Yeah, I am going to remove it from all the other places in the same test scala class

ijuma commented on docs/upgrade.html:
```diff
@@ -97,6 +97,11 @@ <h5><a id="upgrade_200_notable" href="#upgrade_200_notable">Notable changes in 2
+    <li><a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-176%3A+Remove+deprecated+new-consumer+option+for+tools">KIP-176</a> removes
+        the <code>--new-consumer</code> option for all consumer based tools as <code>kafka-console-consumer</code>, <code>kafka-consumer-perf-test</code>
+        and <code>kafka-consumer-groups</code>. The new consumer is automatically used if the bootstrap servers list is provided on the command line
+        otherwise, providing the zookeeper connection instead, the old one is used.
```
We should make it clear that we haven't changed the behaviour, `-new-consumer` was already a no-op.

ppatierno committed:
Addressed some reequested changes on review

ppatierno commented:
@ijuma thanks for the review. I addressed the change requests.

ppatierno committed:
Added missing field in the ConsumerPerformanceConfig (lost after a rebase)

ijuma commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -363,6 +355,8 @@ object ConsumerPerformance extends LazyLogging {
     val hideHeader = options.has(hideHeaderOpt)
     val recordFetchTimeoutMs = options.valueOf(recordFetchTimeoutOpt).longValue()
+    val bootstrapServers = options.valueOf(bootstrapServersOpt)
+    val zkConnectionStr = options.valueOf(zkConnectOpt)
```
These seem to be unused?

vahidhashemian replied:
They are used in the added unit tests in [`ConsumerPerformanceTest.scala`](https://github.com/apache/kafka/pull/5097/files#diff-7e297244dfeb67e4a86f095470b6fa92).

ppatierno replied:
Yes exactly. They are there just for testing that without the - - new-consumer option all depends on the bootstrap servers list and zookeeper connection string and that the provided value to the configuration are what we expect.

ijuma replied:
I see. This is a bit odd and I'd rather not do that. I will push a commit with minor fixes.

vahidhashemian commented on core/src/test/scala/unit/kafka/admin/DeleteConsumerGroupsTest.scala:
```diff
@@ -16,6 +16,7 @@
  */
 package unit.kafka.admin
 
+import joptsimple.OptionException
```
With this import, `joptsimple.` can be dropped from the expected exception class of the first method below.
Also, no need for the `fail` statement there (similar to @ijuma's earlier comment). Not your code, but it would be great to clean them up :)

ppatierno replied:
Yes it makes sense. I have done the same for the `testConfigUsingNewConsumerUnrecognizedOption` in the `ConsumerPerformanceTest` class.

vahidhashemian replied:
Perfect. Thanks!

ppatierno committed:
Added import package and removed useless fail

ijuma committed:
Minor tweaks

ijuma reviewed (approved):
LGTM. I agree that we probably don't want to keep the tests for --new-consumer for long, but we can include them for now and remove them in one or two release cycles.

ijuma commented:
Test failure is unrelated, merging to trunk.
